### ISSUE TITLE
AnimationTreeSync issues with Blend2,Add2,Sub2 style nodes

### ISSUE THREAD

**Issue by catprisbrey** [maji-git/multiplay-core#10]
In my rounds of testing the anim_tree branch of improvements. Currently the following nodes work great under a BlendTree:

- OneShot
- Transition
- TimeScale
- StateMachine (use of advance expressions confirmed working)
- BlendSpace1D
- BlendSpace2D

Tested and unstable nodes:

- Blend2
- Blend3
- Sub2
- Add2

**The Setup:**
These nodes take in a source animation, and then mix/add/sub that animation with a 2nd animation. They seem to sync to peers in a strange ways, but they play normally on whatever instance is playing them locally. 

Typically a blend will either be a lerped value, or a set float value of 0.0 to 1.0, (or -1.0 to 1.0).
Mostly common use cases is to filter bones with it. For example, blending an aiming arm position, on top of an idle or walk animation, and filtering only the arm bones so any movement animation will play but with the arms in an aiming position.

Or blending eye/mouth bone animations on top of any other animation. 

_**During testing this issue is most obvious to see if you have some bone filter turned for comparing what's not syncing.**_

**The Issue:**

If lerping the blend/add amount, the animation blend stutters, trying to blend or add the 2nd animation, and once fully blended to the target amount, the animation for peers, is only the source animation, and not the 2nd blended animation. If not lerping, and instead, a simple set value to the blend amount, on peers, they'll only see 1 frame of the blend, and then only the source animation.

However the blend style nodes work, they don't work well for syncing currently since that source animation currently clobbers the 2nd animation.

**Comment by maji-git** [comment]
I felt like lerp could be out of context, yes I know it's commonly used in animation trees. But developers do have different methods of implementing their lerp function, felt like it should be up to the developers to synchronize it manually on each clients. Feel free to correct me on this too, I'm not animation tree guy :P

also, I'm still not really sure about the issues on Sub2 and Add2, Could you explain more about this if you don't mind?

**Comment by maji-git** [comment]
> I felt like lerp could be out of context, yes I know it's commonly used in animation trees. But developers do have different methods of implementing their lerp function, felt like it should be up to the developers to synchronize it manually on each clients. Feel free to correct me on this too, I'm not animation tree guy :P
> 
> also, I'm still not really sure about the issues on Sub2 and Add2, Could you explain more about this if you don't mind?

And also. If this is really the case, then the only thing multiplay could do is probably ignoring property related to lerp.

**Comment by catprisbrey** [comment]
I can provide a video or something to show the issue, but lerp isn't the issue sadly. As i said before as well, "_a simple set value to the blend amount, on peers, they'll only see 1 frame of the blend, and then only the source animation."_ 

If if you don't lerp at all, and just set the blend value, for some reason on blend2/3, sub2, and add2, it will only show that first frame of the blend, and then revert back to syncing only the source animation. Lerping actually helps because then it's updating that blend every frame, so at least the blend of the two animations works properly for the course of the lerp (with some stuttger) but then once the lerp finishes, (and it's acting like a normal set(parameter/blend2/blend_amount), the source animation clobbers the blend, and you only see the source animation, none of the blend animation.

I'm mid jam at the moment but can swing back to this in a few days.

**Comment by maji-git** [comment]
ooh aight. yea a video, or minimal demo project would help a lot!

also take your time, knowing you're working on your own game tho. best wishes! 🎈

**Comment by catprisbrey** [comment]
[BlendMinimal.zip](https://github.com/maji-git/multiplay-core/files/15380861/BlendMinimal.zip)


Here is a minimal project.
The platform pillars show 3 majis. Left maji is a pose loop, right maji is a walk loop, and the middle maji every 3 seconds will turn flip the blend amount between 0 and 1. That middle maji is how blend2s are supposed to work, blending one animation onto another, and the blend amount stays constant.

The players that spawn have the same animation tree blends, and by pressing space bar you can set the blend amount from 0 to 1 in the exact same way that the middle maji does. But you'll see two weird things happen.

1. For the host, the blend amount after one animation cycle drops back to 0, it doesn't stay constant on the new target amount. (could keyboard input echos be the reason for that?)
2. For multiplayer friends, they only see one frame of this animation from other players.

The Label3D's on each player will make it easy to see how they flip from 0 to 1 for only an instant when multiplayer is running.

Hopefully this helps.

**Comment by maji-git** [comment]
@catprisbrey I can't seem to find the keyboard handling code, where was it exactly?

Simple Blend2 synchronizing without keyboard input works. (also catch a bug along the way lmo)

**Comment by catprisbrey** [comment]
The script for the animation tree.

I usually put it on the character body. But since this is just minimum
example, and the issue is with the animation tree's blend nodes, I decided
to just add it to the animation tree instead. I don't know why. Not my
usual.

You'll see the _input function there.

**Comment by maji-git** [comment]
Alright, I can see the issue now. The issue could be from that input script. The input wasn't checked by MPPlayer, that's why it's modifying the value for everyone. Testing the issue with timer instead of user input do sync properly. confirming this

**Comment by maji-git** [comment]
Alright confirmed.

There's a problem of the demo is that it was not checking from MultiPlay Input. Instead checking from Godot's Input directly, that's what likely caused the stuttering issue. 

However I've spotted critical sync condition issue, and is now addressed in https://github.com/maji-git/multiplay-core/commit/acb0d060c5518e4873fec29e17d01625f03ad322

It should work now if you use MultiPlay Input with the fix on the master branch. 🚀
